### PR TITLE
Add `dolt_foreign_key_status` table function.

### PR DESCRIPTION
**Pull request by PavelSafronov**
`dolt_foreign_key_status` will return the status/information for a foreign key, including its resolved status and the names of the columns involved.
This is a new table function that provides us with the information we need for `dolt diff`.

### CODE REVIEW

**Comment by fulghum** [comment]
Any chance of being able to pull this information from `information_schema` instead? I think the `is_resolved` bit is probably the only thing that wouldn't be there. How badly do we need to expose that bit to customers? Seems like it'd be nice to keep `is_resolved` internal so customers don't need to know about that detail, but maybe we need it for the diff?

**Comment by PavelSafronov** [comment]
> Any chance of being able to pull this information from `information_schema` instead? I think the `is_resolved` bit is probably the only thing that wouldn't be there. How badly do we need to expose that bit to customers? Seems like it'd be nice to keep `is_resolved` internal so customers don't need to know about that detail, but maybe we need it for the diff?

You're right that most of this information is accessible from `information_schema`, but for `dolt diff` we do need to know if a FK is resolved or not, and this information is definitely not in SQL. Daylon suggested this very function:
image

**Comment by nicktobey** [comment]
This definitely needs to be documented in https://docs.dolthub.com/sql-reference/version-control/dolt-sql-functions

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 1]
```diff
@@ -4906,3 +4906,100 @@ END`,
 		},
 	},
 }
+
+var DoltForeignKeyStatus = []queries.ScriptTest{
+	{
+		Name: "Show unresolved foreign key status",
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries.go:4912]: You should also have tests that check for the different failure conditions in your implementation, like calling the function with inputs of the wrong type, or with unresolvable references, etc.

File: go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	return rows, nil
+}
+
+func addColToAllRows(rows []sql.Row, colIndex int, data ...string) []sql.Row {
+	newRows := []sql.Row{}
+	for _, datum := range data {
+		for _, row := range rows {
+			newRow := row.Copy()
+			newRow[colIndex] = datum
+			newRows = append(newRows, newRow)
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go:297]: I don't think you need to make a new row array each time;  just pass in a pointer to the row array and modifying the rows.

**Review by nicktobey — COMMENTED** [review]
Looks good, just a couple of comments about tests and documentation.

File: go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+		ctx:      ctx,
+		database: db,
+	}
+
+	node, err := newInstance.WithExpressions(expressions...)
+	if err != nil {
+		return nil, err
+	}
+
+	ds = node.(*ForeignKeyStatusTableFunction)
```
[the remark this answers is not in the document]
Comment (PavelSafronov) [reply go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go:66]: Removed the cast, returning `node` now.

File: go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go (hunk 2 above)
Comment (PavelSafronov) [reply go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go:297]: We do want to make copies of rows.
The idea here is that we need to return a list of fields for one of the columns.
We're doing this for multiple columns, but let's just consider `table_column`. So if we have 2 values for this column, "foo" and "bar", we want to produce 2 almost-identical  rows of values: (a,b,c,foo,d) and (a,b,c,bar,d).

**Comment by PavelSafronov** [comment]
After offline discussion, it was decided that foreign key status (resolved vs not resolved) should not be surfaced to users, so I'm closing this PR without merging.

**Pull request closed without merging**